File: app/src/main/java/my/music/notes/ExampleAdapter.java
package my.music.notes;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class ExampleAdapter extends RecyclerView.Adapter<ExampleAdapter.ExampleViewHolder> {

    private ArrayList<ExampleItem> exampleItems;


    public ExampleAdapter(ArrayList<ExampleItem> exampleItems){
        this.exampleItems = exampleItems;
    }

    @NonNull
    @Override
    public ExampleViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.example_item, parent, false);

        return new ExampleViewHolder(v, listener);
    }

    @Override
    public void onBindViewHolder(@NonNull ExampleViewHolder holder, int position) {
        ExampleItem currentItem = this.exampleItems.get(position);

        holder.imageView.setImageResource(currentItem.getImageResource());
        holder.textView1.setText(currentItem.getText1());
        holder.textView2.setText(currentItem.getText2());
    }

    @Override
    public int getItemCount() {
        return exampleItems.size();
    }

    private OnItemClickListener listener;

    public interface OnItemClickListener{
        void OnItemClick(int position);
        void OnDeleteClick(int position);
    }

    public void setOnclickListener(OnItemClickListener listener){
        this.listener = listener;
    }
    public static class ExampleViewHolder extends RecyclerView.ViewHolder{

        public ImageView imageView;
        public TextView textView1;
        public TextView textView2;
        public ImageView imageItemDelete;

        public ExampleViewHolder(@NonNull View itemView, final OnItemClickListener listener) {
            super(itemView);

            this.imageView = itemView.findViewById(R.id.imageView);
            this.textView1 = itemView.findViewById(R.id.textView1);
            this.textView2 = itemView.findViewById(R.id.textView2);
            this.imageItemDelete = itemView.findViewById(R.id.image_item_delete);

            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    if (listener != null){
                        int position = getAdapterPosition();
                        if (position != RecyclerView.NO_POSITION){
                            listener.OnItemClick(position);
                        }
                    }
                }
            });

            imageItemDelete.setOnClickListener(
                    new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            if (listener != null){
                                int position = getAdapterPosition();
                                if (position != RecyclerView.NO_POSITION){
                                    listener.OnDeleteClick(position);
                                }
                            }
                        }
                    });
        }


    }
}

File: app/src/main/java/my/music/notes/adapters/ExampleAdapter.kt
package my.music.notes.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.example_item.view.*
import my.music.notes.R
import my.music.notes.models.ExampleItem

// ListAdapter out of the box update list async on each submitList + nice animation
class ExampleAdapter(
        private val onItemClicked: (ExampleItem, Int) -> Unit,
        private val onItemDeleteClicked: (ExampleItem, Int) -> Unit
) : ListAdapter<ExampleItem, ExampleAdapter.ExampleViewHolder>(diffUtilItemCallback) {

    companion object {
        private val diffUtilItemCallback = object : DiffUtil.ItemCallback<ExampleItem>() {
            override fun areItemsTheSame(oldItem: ExampleItem, newItem: ExampleItem): Boolean {
                return oldItem.uuid == newItem.uuid //todo add some id's?
            }

            override fun areContentsTheSame(oldItem: ExampleItem, newItem: ExampleItem): Boolean {
                return oldItem == newItem
            }

        }
    }

    inner class ExampleViewHolder(item: View) : RecyclerView.ViewHolder(item) {

        fun bind(model: ExampleItem) {
            itemView.imageView.setImageResource(model.imageResource)
            itemView.textView1.text = model.firstText
            itemView.textView2.text = model.secondaryText
            itemView.image_item_delete.setOnClickListener {
                if (adapterPosition == RecyclerView.NO_POSITION) return@setOnClickListener
                onItemDeleteClicked(model, adapterPosition)
            }
            itemView.setOnClickListener {
                if (adapterPosition == RecyclerView.NO_POSITION) return@setOnClickListener
                onItemClicked(model, adapterPosition)
            }
        }

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ExampleViewHolder {
        return ExampleViewHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.example_item, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ExampleViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

}